File: stryker4jvm-core/src/main/java/stryker4jvm/core/model/LanguageMutator.java
package stryker4jvm.core.model;

import java.nio.file.Path;
import java.util.List;
import java.util.Map;

import stryker4jvm.core.exception.Stryker4jvmException;

/**
 * Class that encapsulates the three components of a mutator for a specific (programming) language.
 *
 * @param <T> The language-specific AST.
 */
public class LanguageMutator<T extends AST> {
  private final Parser<T> parser;
  private final Collector<T> collector;
  private final Instrumenter<T> instrumenter;

  public LanguageMutator(Parser<T> parser, Collector<T> collector, Instrumenter<T> instrumenter) {
    this.parser = parser;
    this.collector = collector;
    this.instrumenter = instrumenter;
  }

  public T parse(Path p) throws Stryker4jvmException {
    return parser.parse(p);
  }

  public CollectedMutants<T> collect(AST tree) {
    return collector.collect((T) tree);
  }

  public T instrument(AST source, Map<AST, List<MutantWithId<AST>>> mutations) {
    return instrumenter.instrument(
        (T) source, (Map<T, List<MutantWithId<T>>>) (Map<T, ?>) mutations);
  }
}

File: stryker4jvm-core/src/main/scala/stryker4jvm/model/LanguageMutator.scala
package stryker4jvm.model

import stryker4jvm.core.model.{AST, CollectedMutants, Collector, Instrumenter, MutantWithId, Parser}
import stryker4jvm.core

import java.nio.file.Path
import java.util

class LanguageMutator[T <: AST](parser: Parser[T], collector: Collector[T], instrumenter: Instrumenter[T]) {
  type Tree = T

  def parse(p: Path): T = parser.parse(p)

  def collect(tree: AST): CollectedMutants[T] = {
    collector.collect(tree.asInstanceOf[T])
  }

  def instrument(source: AST, mutations: util.Map[AST, util.List[MutantWithId[AST]]]): T = {
    instrumenter.instrument(
      source.asInstanceOf[T],
      mutations.asInstanceOf[util.Map[T, util.List[core.model.MutantWithId[T]]]]
    )
  }
}
